Stop decryption when maintenance mode is enabled

RubenHoms opened the pull request:
See discussion in #8311. This should at least make sure that people don't try and decrypt their files and then find out they are still encrypted and can't be opened anymore.

RubenHoms committed:
Stop decryption when maintenance mode is enabled, fixes #8311

Signed-off-by: Ruben Homs <ruben@homs.codes>

MorrisJobke commented on core/Command/Encryption/DecryptAll.php:
```diff
@@ -132,6 +132,13 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+			$output->writeln("Maintenance mode should be disabled during decryption.");
+			$output->writeln("Please disable maintenance mode and try again.");
+			return;
+		}
```
See `resetMaintenanceAndTrashbin` - I guess it was intended that maintenance mode could be enabled during the decryption to avoid parallel writing to files.

cc @schiessle

MorrisJobke replied:
I'm not an expert in the server side encryption thus I mentioned @schiessle to have a look.

RubenHoms replied:
It looks like the trashbin app is disabled and maintenance enabled further down on line 167, so I don't think this will need to call the `resetMaintenanceAndTrashbin`.

MorrisJobke replied:
I was just about to say, that is was designed with "maintenance mode could be enabled" - maybe there is an issue somewhere else that just has the symptoms of "once maintenance mode is off it works".

RubenHoms commented:
I can't explain it much better than [this comment](https://github.com/nextcloud/server/issues/8311#issuecomment-380555487) does.

nickvergessen commented:
Ah, well it's basically quite simple:

```
$ ./occ encryption:decrypt-all 
Nextcloud is in maintenance mode - no apps have been loaded

Disable server side encryption... done.


You are about to start to decrypt all files stored in your Nextcloud.
It will depend on the encryption module and your setup if this is possible.
Depending on the number and size of your files this can take some time
Please make sure that no user access his files during this process!

Do you really want to continue? (y/n) y
```

The first line is the most important one. When the instance is in maintenance mode, no apps are being loaded. Easiest way to verify this is:
```
$ ./occ app:enable encryption
Nextcloud is in maintenance mode - no apps have been loaded

encryption enabled
$ ./occ encryption:list-modules
Nextcloud is in maintenance mode - no apps have been loaded
$
```

So yeah, makes total sense to not allow to execute those commands when in maintenance mode. So I guess the recommended way would be to leave maintenance mode off and instead stop the apache/nginx/web server.

nickvergessen commented on core/Command/Encryption/DecryptAll.php:
```diff
@@ -132,6 +132,13 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 
+		$isMaintenanceModeEnabled = $this->config->getSystemValue('maintenance', false);
+		if($isMaintenanceModeEnabled) {
+			$output->writeln("Maintenance mode should be disabled during decryption.");
```
```
Maintenance mode must be disabled when starting decryption,
in order to load the relevant encryption modules correctly.
Your instance will automatically be put to maintenance mode
during the actual dencryption of the files.
```

nickvergessen replied:
Well list-modules will post an empty list, which might be confusing,
set-default-module doesn't work, apart from setting it to the same module
encrypt-all doesnt work

So at least those.

nickvergessen reviewed (changes requested):
Please also add similar messages to the other `encryption:*` commands to avoid confusion.

RubenHoms commented:
Sorry for the delay, I've added the notice to the `encryption:list-modules` and `encryption:set-default-module` commands as requested by @nickvergessen

RubenHoms commented:
Could someone check this out so we can stop other people from losing data?

RubenHoms commented:
Please, for the sake of people losing their data (like happened to me, also see #8311 for more instances of this happening). Could someone merge this or shoot at it?
@rullzer @nickvergessen @schiessle @kesselb

kesselb commented on core/Command/Encryption/DecryptAll.php:
```diff
@@ -132,6 +132,15 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+			$output->writeln("Maintenance mode must be disabled when starting decryption,");
+			$output->writeln("in order to load the relevant encryption modules correctly.");
+			$output->writeln("Your instance will automatically be put to maintenance mode");
+			$output->writeln("during the actual dencryption of the files.");
```
```suggestion
			$output->writeln("during the actual decryption of the files.");
```

kesselb commented on core/Command/Encryption/DecryptAll.php:
```diff
@@ -132,6 +132,15 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		}
 
+		$isMaintenanceModeEnabled = $this->config->getSystemValue('maintenance', false);
+		if($isMaintenanceModeEnabled) {
```
```suggestion
		if ($isMaintenanceModeEnabled) {
```

kesselb commented on core/Command/Encryption/ListModules.php:
```diff
@@ -49,6 +58,13 @@ protected function configure() {
 
 	protected function execute(InputInterface $input, OutputInterface $output) {
+		$isMaintenanceModeEnabled = $this->config->getSystemValue('maintenance', false);
+		if($isMaintenanceModeEnabled) {
```
```suggestion
		if ($isMaintenanceModeEnabled) {
```

kesselb commented on core/Command/Encryption/SetDefaultModule.php:
```diff
@@ -56,6 +65,13 @@ protected function configure() {
 
 	protected function execute(InputInterface $input, OutputInterface $output) {
+		$isMaintenanceModeEnabled = $this->config->getSystemValue('maintenance', false);
+		if($isMaintenanceModeEnabled) {
```
```suggestion
		if ($isMaintenanceModeEnabled) {
```

kesselb commented:
Could you rebase your branch to latest master so the tests run again?

RubenHoms commented:
@kesselb Thanks for looking into this. I accepted your suggestions, rebased against upstream and squashed it into 1 commit. Now let's see what the tests do.

RubenHoms commented:
@kesselb Tests have failed yet I have no idea whatsoever why. The tests that have failed are about connecting to MySQL/Redis, I don't see how the changes in this PR have anything to do with that. Can someone take a look?

MorrisJobke commented:
> @kesselb Tests have failed yet I have no idea whatsoever why. The tests that have failed are about connecting to MySQL/Redis, I don't see how the changes in this PR have anything to do with that. Can someone take a look?


The constructor arguments have been changed and thus the tests don't properly inject the correct constructor arguments. Check out the test files listed in the output and adapt them to the updated code you added. Then the tests should pass.

```
132     1) Tests\Core\Command\Encryption\SetDefaultModuleTest::testSetDefaultModule with data set #0 ('ID0', 'ID0', null, null, 'already')
133	ArgumentCountError: Too few arguments to function OC\Core\Command\Encryption\SetDefaultModule::__construct(), 1 passed in /drone/src/tests/Core/Command/Encryption/SetDefaultModuleTest.php on line 52 and exactly 2 expected
134	
135	/drone/src/core/Command/Encryption/SetDefaultModule.php:44
136	/drone/src/tests/Core/Command/Encryption/SetDefaultModuleTest.php:52
137	
138	2) Tests\Core\Command\Encryption\SetDefaultModuleTest::testSetDefaultModule with data set #1 ('ID0', 'ID1', 'ID1', true, 'info')
139	ArgumentCountError: Too few arguments to function OC\Core\Command\Encryption\SetDefaultModule::__construct(), 1 passed in /drone/src/tests/Core/Command/Encryption/SetDefaultModuleTest.php on line 52 and exactly 2 expected
140	
141	/drone/src/core/Command/Encryption/SetDefaultModule.php:44
142	/drone/src/tests/Core/Command/Encryption/SetDefaultModuleTest.php:52
143	
144	3) Tests\Core\Command\Encryption\SetDefaultModuleTest::testSetDefaultModule with data set #2 ('ID0', 'ID1', 'ID1', false, 'error')
145	ArgumentCountError: Too few arguments to function OC\Core\Command\Encryption\SetDefaultModule::__construct(), 1 passed in /drone/src/tests/Core/Command/Encryption/SetDefaultModuleTest.php on line 52 and exactly 2 expected
146	
147	/drone/src/core/Command/Encryption/SetDefaultModule.php:44
148	/drone/src/tests/Core/Command/Encryption/SetDefaultModuleTest.php:52
```


You can also execute those files locally by calling this:

```
./autotest.sh sqlite tests/Core/Command/Encryption/SetDefaultModuleTest.php
````

kesselb commented:
Thank you @RubenHoms :tada: 

I have some suggestions for the test. Usually we try to mock objects in unit tests. I added a mocked version of IConfig and added a test case that some methods are not called if maintenance mode is enabled. Let me know if you have any questions ;)

```
Index: tests/Core/Command/Encryption/SetDefaultModuleTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/Core/Command/Encryption/SetDefaultModuleTest.php	(revision a3b4410283a0517b458deacc8ee704c12ec028bd)
+++ tests/Core/Command/Encryption/SetDefaultModuleTest.php	(date 1558186444000)
@@ -24,13 +24,16 @@
 
 use OC\Core\Command\Encryption\SetDefaultModule;
 use OCP\Encryption\IManager;
+use OCP\IConfig;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Output\OutputInterface;
 use Test\TestCase;
 
 class SetDefaultModuleTest extends TestCase {
-	/** @var \PHPUnit_Framework_MockObject_MockObject */
+	/** @var \PHPUnit_Framework_MockObject_MockObject|IManager */
 	protected $manager;
+	/** @var \PHPUnit_Framework_MockObject_MockObject|IConfig */
+	protected $config;
 	/** @var \PHPUnit_Framework_MockObject_MockObject */
 	protected $consoleInput;
 	/** @var \PHPUnit_Framework_MockObject_MockObject */
@@ -42,9 +45,12 @@
 	protected function setUp() {
 		parent::setUp();
 
-		$manager = $this->manager = $this->getMockBuilder(IManager::class)
+		$this->manager = $this->getMockBuilder(IManager::class)
 			->disableOriginalConstructor()
 			->getMock();
+		$this->config = $this->getMockBuilder(IConfig::class)
+			->getMock();
+
 		$this->consoleInput = $this->getMockBuilder(InputInterface::class)->getMock();
 		$this->consoleOutput = $this->getMockBuilder(OutputInterface::class)->getMock();
 
@@ -79,6 +85,12 @@
 		$this->manager->expects($this->once())
 			->method('getDefaultEncryptionModuleId')
 			->willReturn($oldModule);
+
+		$this->config->expects($this->once())
+			->method('getSystemValue')
+			->with('maintenance', false)
+			->willReturn(false);
+
 		if ($updateModule) {
 			$this->manager->expects($this->once())
 				->method('setDefaultEncryptionModule')
@@ -90,6 +102,41 @@
 			->method('writeln')
 			->with($this->stringContains($expectedString));
 
+		self::invokePrivate($this->command, 'execute', [$this->consoleInput, $this->consoleOutput]);
+	}
+
+	/**
+	 * @dataProvider dataSetDefaultModule
+	 *
+	 * @param string $oldModule
+	 * @param string $newModule
+	 * @param string $updateModule
+	 * @param bool $updateSuccess
+	 * @param string $expectedString
+	 */
+	public function testMaintenanceMode($oldModule, $newModule, $updateModule, $updateSuccess, $expectedString) {
+		$this->consoleInput->expects($this->never())
+			->method('getArgument')
+			->with('module')
+			->willReturn($newModule);
+
+		$this->manager->expects($this->never())
+			->method('getDefaultEncryptionModuleId')
+			->willReturn($oldModule);
+
+		$this->config->expects($this->once())
+			->method('getSystemValue')
+			->with('maintenance', false)
+			->willReturn(true);
+
+		$this->consoleOutput->expects($this->at(0))
+			->method('writeln')
+			->with($this->stringContains('Maintenance mode must be disabled when setting default module,'));
+
+		$this->consoleOutput->expects($this->at(1))
+			->method('writeln')
+			->with($this->stringContains('in order to load the relevant encryption modules correctly.'));
+
 		self::invokePrivate($this->command, 'execute', [$this->consoleInput, $this->consoleOutput]);
 	}
 }

```

kesselb commented:
Looks very good :+1: 

I made a small mistake :see_no_evil: The `setup` method should look like this:
```

	protected function setUp() {
		parent::setUp();

		$this->manager = $this->getMockBuilder(IManager::class)
			->disableOriginalConstructor()
			->getMock();
		$this->config = $this->getMockBuilder(IConfig::class)
			->getMock();

		$this->consoleInput = $this->getMockBuilder(InputInterface::class)->getMock();
		$this->consoleOutput = $this->getMockBuilder(OutputInterface::class)->getMock();

		$this->command = new SetDefaultModule($this->manager, $this->config);
	}
```

kesselb commented:
Could you rebase, signoff the commits and than we are good to go ;)

RubenHoms commented:
@kesselb thanks for the code for the tests. I'm not familiar with the test suite itself and I can't make sense out of the tests that are still failing. Any input to fix this would be appreciated.

kesselb commented:
Thank you :tada: 

I'm not 100% sure but changing the expects at should be enough. We introduced another call (is maintenance true/false) and the other calls are +1.

```
Index: tests/Core/Command/Encryption/DecryptAllTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/Core/Command/Encryption/DecryptAllTest.php	(revision ae7f89fd9f2bb87c0254f5fa087d0d9472612644)
+++ tests/Core/Command/Encryption/DecryptAllTest.php	(date 1558364995000)
@@ -148,7 +148,7 @@
 			->willReturn('user1');
 
 		if ($encryptionEnabled) {
-			$this->config->expects($this->at(0))
+			$this->config->expects($this->at(1))
 				->method('setAppValue')
 				->with('core', 'encryption_enabled', 'no');
 			$this->questionHelper->expects($this->once())
@@ -160,7 +160,7 @@
 					->with($this->consoleInput, $this->consoleOutput, 'user1');
 			} else {
 				$this->decryptAll->expects($this->never())->method('decryptAll');
-				$this->config->expects($this->at(1))
+				$this->config->expects($this->at(2))
 					->method('setAppValue')
 					->with('core', 'encryption_enabled', 'yes');
 			}
@@ -194,12 +194,12 @@
 			$this->questionHelper
 		);
 
-		$this->config->expects($this->at(0))
+		$this->config->expects($this->at(1))
 			->method('setAppValue')
 			->with('core', 'encryption_enabled', 'no');
 
 		// make sure that we enable encryption again after a exception was thrown
-		$this->config->expects($this->at(3))
+		$this->config->expects($this->at(4))
 			->method('setAppValue')
 			->with('core', 'encryption_enabled', 'yes');
```

cc @nextcloud/encryption

RubenHoms commented:
@kesselb I applied your patch and let the tests run. The two that are failing now seem to be failing in all other tests that are being run (`MySQL has gone away`) so I don't think that has anything to do with the introduction of this change.

MorrisJobke commented:
> @kesselb I applied your patch and let the tests run. The two that are failing now seem to be failing in all other tests that are being run (`MySQL has gone away`) so I don't think that has anything to do with the introduction of this change.

The tests are fine - there is just sometimes a hiccup where the unit tests run super slow and then the timeout hits in and kills the mysql first causing the error in the unit tests. So all looks good now 👍

MorrisJobke reviewed (approved):
Code makes sense 👍

MorrisJobke commented:
@rullzer I removed the back port request. Or do you still want this?

FlorianFranzen commented:
@RubenHoms Thank you a lot for following throught with this!

RubenHoms commented:
Happy to help out. :ok_hand: Thanks for all of your input.
